kernel/fatal.c

```diff
@@ -0,0 +1,92 @@
+#include <kernel.h>
+#include <kernel_internal.h>
+#include <misc/printk.h>
+#include <arch/cpu.h>
+#include <logging/log_ctrl.h>
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
+{
+	ARG_UNUSED(reason);
+
+	for (;;) {
+		k_cpu_idle();
+	}
+}
+/* LCOV_EXCL_STOP */
+
+/* LCOV_EXCL_START */
+__weak void k_sys_fatal_error_handler(unsigned int reason,
+				      const NANO_ESF *esf)
+{
+	ARG_UNUSED(esf);
+
+	printk("Halting system.\n");
+	(void)irq_lock();
```

Comment:
This should be z_arch_irq_lock().  On SMP, irq_lock() is a spinlock with special semantics.  Also see note elsewhere, I think this wants to be part of the halt() implementation.

Reply:
Even on SMP systems, `irq_lock()` will give us the behavior we want, no?  As soon as the other CPUs attempt to grab the global lock they'll just hang.  We could nudge the process along with an IPI but I don't see that this needs changing.

Reply:
for now I have changed to z_arch_irq_lock(), but I think @gnuless makes a good point. Leaving this open until we decide the best course of action.

Reply:
Sorry, missed this comment yesterday: it's true that irq_lock() will work to suspend local interrupts, so it will work for the task desired (at the cost of including the code for it which might otherwise be elided if the system is fully spinlockified).  But it won't work as a replacement for a portable halt of other CPUs.

That idea would work if irq_lock() was the universally used global lock, but it's not anymore.  The core kernel synchronization has been split into a bunch of independent spinlocks (which is a feature, obviously) and so there's no guarantee that anything is going to hit this at any point.  You'd need to be using a driver that happened to be using the legacy API.

Reply:
I assume the point of the lock is that ISRs will grab it upon entry, no?